File: Javascript/Arrays/subarraySort.js
// Here is your solution
function subarraySort(array) {
  var maxOutOfOrder = -Infinity;
  var minOutOfOrder = Infinity;
  for (var idx = 0; idx < array.length; idx++) {
    var num = array[idx];
    if (isOutOfOrder(idx, num, array)) {
      maxOutOfOrder = Math.max(maxOutOfOrder, num);
      minOutOfOrder = Math.min(minOutOfOrder, num);
    }
  }
  if (minOutOfOrder === Infinity) {
    return [-1, -1];
  }
  var leftOutOfOrderIdx = 0;
  while (minOutOfOrder >= array[leftOutOfOrderIdx]) {
    leftOutOfOrderIdx++;
  }
  var rightOutOfOrder = array.length - 1;
  while (maxOutOfOrder < array[rightOutOfOrder]) {
    rightOutOfOrder--;
  }
  return [leftOutOfOrderIdx, rightOutOfOrder];
}

function isOutOfOrder(idx, num, array) {
  if (idx == 0) {
    return num > array[idx + 1];
  }
  if (idx == array.length - 1) {
    return num < array[array.length - 1];
  }
  return num < array[idx - 1] || num > array[idx + 1];
}
// Here is your Output
var array = [1, 2, 4, 7, 10, 11, 7, 12, 6, 7, 16, 18, 19];
var output = subarraySort(array);
console.log("👋 Output:", output);

File: Python/Arrays/subarraySort.py
# O(n) time | O(1) space
def subarraySort(array):
    maxOutOfOrder = float("-inf")
    minOutOfOrder = float("inf")
    for idx in range(len(array)):
        num = array[idx]
        if isOutOfOrder(idx, num, array):
            maxOutOfOrder = max(maxOutOfOrder, num)
            minOutOfOrder = min(minOutOfOrder, num)

    if minOutOfOrder == float("inf"):
        return [-1, -1]

    leftOutOfOrderIdx = 0
    rightOutOfOrder = len(array) - 1

    while (minOutOfOrder >= array[leftOutOfOrderIdx]):
        leftOutOfOrderIdx += 1

    while (maxOutOfOrder <= array[rightOutOfOrder]):
        rightOutOfOrder -= 1

    return [leftOutOfOrderIdx, rightOutOfOrder]

    print(maxOutOfOrder)
    print(minOutOfOrder)

    return array

def isOutOfOrder(idx, num, array):
    if idx == 0:
        return num > array[idx + 1]
    if idx == len(array) - 1:
        return num < array[idx - 1]
    return num > array[idx + 1] or num < array[idx - 1] 

# Here is your Output
array = [1, 2, 4, 7, 10, 11, 7, 12, 6, 7, 16, 18, 19]
output = subarraySort(array)
print("👋 Output:", output);

# [1, 2, 4, 7, 10, 11, 7, 12, 6, 7, 16, 18, 19]
#  1  2  3  4   5   6  7   8  9  10 11  12  13
#  1  2  4  6   7   7  7  10  11 12 16  18  19
#  